File: app/services/product_profiler.py
import json
from urllib.parse import urlparse

import httpx
from bs4 import BeautifulSoup

from app.config import Settings
from app.schemas import ProductProfile, ProductProfileOverrides
from app.utils import host_to_store_name, normalize_url, normalize_whitespace, unique_preserve_order


class ProductProfiler:
    USER_AGENT = (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36"
    )

    def __init__(self, settings: Settings):
        self.settings = settings

    def build_profile(self, product_url: str, overrides: ProductProfileOverrides | None = None) -> ProductProfile:
        overrides = overrides or ProductProfileOverrides()
        headers = {"User-Agent": self.USER_AGENT, "Accept-Language": "es-AR,es;q=0.9,en;q=0.8"}
        with httpx.Client(timeout=self.settings.request_timeout_seconds, follow_redirects=True, headers=headers) as client:
            response = client.get(product_url)
            response.raise_for_status()

        final_url = str(response.url)
        soup = BeautifulSoup(response.text, "lxml")
        parsed_url = urlparse(final_url)
        domain = parsed_url.netloc.lower()
        notes: list[str] = []

        page_title = normalize_whitespace(soup.title.string if soup.title and soup.title.string else "") or None
        h1 = normalize_whitespace(soup.find("h1").get_text(" ", strip=True) if soup.find("h1") else "") or None
        canonical_url = self._extract_canonical_url(soup) or final_url
        meta_description = self._extract_meta_content(soup, "description") or self._extract_meta_property(soup, "og:description")
        og_title = self._extract_meta_property(soup, "og:title")

        product_schema = self._extract_schema_object(soup, "Product")
        organization_schema = self._extract_schema_object(soup, "Organization")

        product_name = self._first_non_empty(
            overrides.product_name,
            self._schema_value(product_schema, "name"),
            og_title,
            h1,
            page_title,
        )
        if not product_name:
            raise RuntimeError("No se pudo extraer un nombre de producto util desde la URL")

        brand_name = self._first_non_empty(
            overrides.brand_name,
            self._extract_brand_name(product_schema),
            self._extract_meta_property(soup, "product:brand"),
        )
        store_name = self._first_non_empty(
            overrides.store_name,
            self._schema_value(organization_schema, "name"),
            host_to_store_name(domain),
        )
        category = self._first_non_empty(
            overrides.category,
            self._schema_value(product_schema, "category"),
            self._extract_breadcrumb_category(soup),
        )

        aliases = self._build_aliases(product_name, brand_name, overrides.aliases)
        vendor_aliases = unique_preserve_order(overrides.vendor_aliases + ([store_name] if store_name else []) + ([brand_name] if brand_name else []))
        competitor_names = unique_preserve_order(overrides.competitor_names)

        if product_schema:
            notes.append("Product schema detected")
        if organization_schema:
            notes.append("Organization schema detected")
        if overrides.model_dump(exclude_none=True, exclude_defaults=True):
            notes.append("Manual overrides applied")

        return ProductProfile(
            source_url=product_url,
            canonical_url=normalize_url(overrides.canonical_url or canonical_url),
            domain=domain,
            product_name=normalize_whitespace(product_name),
            brand_name=normalize_whitespace(brand_name) or None,
            store_name=normalize_whitespace(store_name) or None,
            category=normalize_whitespace(category) or None,
            page_title=page_title,
            meta_description=normalize_whitespace(meta_description) or None,
            aliases=aliases,
            vendor_aliases=vendor_aliases,
            competitor_names=competitor_names,
            extraction_notes=notes,
        )

    def _extract_canonical_url(self, soup: BeautifulSoup) -> str | None:
        canonical = soup.find("link", rel=lambda value: value and "canonical" in value.lower())
        if canonical and canonical.get("href"):
            return canonical.get("href")
        return None

    def _extract_meta_content(self, soup: BeautifulSoup, name: str) -> str | None:
        tag = soup.find("meta", attrs={"name": name})
        return tag.get("content") if tag and tag.get("content") else None

    def _extract_meta_property(self, soup: BeautifulSoup, prop: str) -> str | None:
        tag = soup.find("meta", attrs={"property": prop})
        return tag.get("content") if tag and tag.get("content") else None

    def _extract_schema_object(self, soup: BeautifulSoup, schema_type: str) -> dict | None:
        for script in soup.find_all("script", attrs={"type": "application/ld+json"}):
            raw = script.string or script.get_text(" ", strip=True)
            if not raw:
                continue
            try:
                payload = json.loads(raw)
            except json.JSONDecodeError:
                continue
            for candidate in self._iter_schema_items(payload):
                candidate_type = candidate.get("@type")
                if isinstance(candidate_type, list) and schema_type in candidate_type:
                    return candidate
                if candidate_type == schema_type:
                    return candidate
        return None

    def _iter_schema_items(self, payload: object):
        if isinstance(payload, list):
            for item in payload:
                yield from self._iter_schema_items(item)
        elif isinstance(payload, dict):
            if "@graph" in payload and isinstance(payload["@graph"], list):
                for item in payload["@graph"]:
                    yield from self._iter_schema_items(item)
            else:
                yield payload

    def _schema_value(self, schema: dict | None, key: str) -> str | None:
        if not schema:
            return None
        value = schema.get(key)
        if isinstance(value, dict):
            return value.get("name") or value.get("value")
        if isinstance(value, str):
            return value
        return None

    def _extract_brand_name(self, product_schema: dict | None) -> str | None:
        if not product_schema:
            return None
        brand = product_schema.get("brand")
        if isinstance(brand, dict):
            return brand.get("name") or brand.get("brand")
        if isinstance(brand, str):
            return brand
        return None

    def _extract_breadcrumb_category(self, soup: BeautifulSoup) -> str | None:
        breadcrumbs = soup.select("nav[aria-label*='breadcrumb'] a, .breadcrumb a")
        items = [normalize_whitespace(node.get_text(" ", strip=True)) for node in breadcrumbs]
        items = [item for item in items if item]
        if len(items) >= 2:
            return items[-2]
        return None

    def _build_aliases(self, product_name: str, brand_name: str | None, manual_aliases: list[str]) -> list[str]:
        candidates = [product_name]
        if brand_name and product_name.lower().startswith(brand_name.lower()):
            candidates.append(product_name[len(brand_name) :].strip(" -"))
        candidates.extend(manual_aliases)
        compact = product_name.replace("-", " ")
        if compact != product_name:
            candidates.append(compact)
        return unique_preserve_order(candidates)

    def _first_non_empty(self, *values: str | None) -> str | None:
        for value in values:
            cleaned = normalize_whitespace(value)
            if cleaned:
                return cleaned
        return None

File: src/lib/product-profiler.ts
import * as cheerio from "cheerio";

import type { ProductProfile, ProductProfileOverrides } from "@/lib/types";
import { hostToStoreName, normalizeUrl, normalizeWhitespace, uniquePreserveOrder } from "@/lib/utils";

export async function buildProductProfile(productUrl: string, overrides: ProductProfileOverrides = {}): Promise<ProductProfile> {
  const response = await fetch(productUrl, {
    headers: {
      "user-agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
      "accept-language": "es-AR,es;q=0.9,en;q=0.8",
    },
    cache: "no-store",
    redirect: "follow",
  });

  if (!response.ok) {
    throw new Error(`No se pudo leer la URL del producto (${response.status})`);
  }

  const html = await response.text();
  const finalUrl = response.url;
  const url = new URL(finalUrl);
  const domain = url.hostname.toLowerCase();
  const $ = cheerio.load(html);
  const extractionNotes: string[] = [];

  const pageTitle = normalizeWhitespace($("title").first().text()) || null;
  const h1 = normalizeWhitespace($("h1").first().text()) || null;
  const canonicalUrl = $("link[rel='canonical']").attr("href") || finalUrl;
  const metaDescription = normalizeWhitespace($("meta[name='description']").attr("content")) || null;
  const ogTitle = normalizeWhitespace($("meta[property='og:title']").attr("content")) || null;

  const productSchema = extractBestProductSchema($, [h1, ogTitle, pageTitle]);
  const organizationSchema = extractSchemaObject($, "Organization");

  const productName = firstNonEmpty(
    overrides.productName,
    h1,
    ogTitle,
    schemaValue(productSchema, "name"),
    pageTitle,
  );

  if (!productName) {
    throw new Error("No se pudo extraer un nombre de producto util desde la URL");
  }

  const brandName = firstNonEmpty(overrides.brandName, extractBrandName(productSchema), $("meta[property='product:brand']").attr("content"));
  const storeName = firstNonEmpty(overrides.storeName, schemaValue(organizationSchema, "name"), hostToStoreName(domain));
  const category = firstNonEmpty(overrides.category, schemaValue(productSchema, "category"), extractBreadcrumbCategory($));

  const aliases = buildAliases(productName, brandName ?? undefined, overrides.aliases ?? []);
  const vendorAliases = uniquePreserveOrder([...(overrides.vendorAliases ?? []), storeName ?? "", brandName ?? ""]);
  const competitorNames = uniquePreserveOrder(overrides.competitorNames ?? []);

  if (productSchema) extractionNotes.push("Product schema detected");
  if (organizationSchema) extractionNotes.push("Organization schema detected");
  if (Object.keys(overrides).length > 0) extractionNotes.push("Manual overrides applied");

  return {
    sourceUrl: productUrl,
    canonicalUrl: normalizeUrl(overrides.canonicalUrl ?? canonicalUrl),
    domain,
    productName,
    brandName: brandName ?? inferBrandFromProductName(productName) ?? null,
    storeName: storeName ?? null,
    category: category ?? null,
    pageTitle,
    metaDescription,
    aliases,
    vendorAliases,
    competitorNames,
    extractionNotes,
  };
}

function extractBestProductSchema($: cheerio.CheerioAPI, references: Array<string | null | undefined>): Record<string, unknown> | null {
  const candidates = extractSchemaObjects($, "Product");
  if (!candidates.length) {
    return null;
  }

  const referenceTokens = tokenizeForMatch(references.join(" "));
  if (!referenceTokens.length) {
    return candidates[0];
  }

  let best = candidates[0];
  let bestScore = -1;
  for (const candidate of candidates) {
    const candidateName = schemaValue(candidate, "name") || "";
    const candidateTokens = tokenizeForMatch(candidateName);
    if (!candidateTokens.length) {
      continue;
    }

    let score = 0;
    for (const token of candidateTokens) {
      if (referenceTokens.includes(token)) {
        score += 1;
      }
    }

    if (score > bestScore) {
      best = candidate;
      bestScore = score;
    }
  }

  return best;
}

function extractSchemaObject($: cheerio.CheerioAPI, schemaType: string): Record<string, unknown> | null {
  const candidates = extractSchemaObjects($, schemaType);
  return candidates[0] ?? null;
}

function extractSchemaObjects($: cheerio.CheerioAPI, schemaType: string): Record<string, unknown>[] {
  const scripts = $("script[type='application/ld+json']")
    .toArray()
    .map((element) => $(element).contents().text())
    .filter(Boolean);

  const candidates: Record<string, unknown>[] = [];
  for (const raw of scripts) {
    try {
      const payload = JSON.parse(raw) as unknown;
      for (const candidate of iterateSchemaItems(payload)) {
        const candidateType = candidate["@type"];
        if (Array.isArray(candidateType) && candidateType.includes(schemaType)) {
          candidates.push(candidate);
        }
        if (candidateType === schemaType) {
          candidates.push(candidate);
        }
      }
    } catch {
      continue;
    }
  }

  return candidates;
}

function* iterateSchemaItems(payload: unknown): Generator<Record<string, unknown>> {
  if (Array.isArray(payload)) {
    for (const item of payload) {
      yield* iterateSchemaItems(item);
    }
    return;
  }

  if (!payload || typeof payload !== "object") {
    return;
  }

  const record = payload as Record<string, unknown>;
  const graph = record["@graph"];
  if (Array.isArray(graph)) {
    for (const item of graph) {
      yield* iterateSchemaItems(item);
    }
    return;
  }

  yield record;
}

function schemaValue(schema: Record<string, unknown> | null, key: string): string | undefined {
  if (!schema) {
    return undefined;
  }

  const value = schema[key];
  if (typeof value === "string") {
    return normalizeWhitespace(value);
  }
  if (value && typeof value === "object") {
    const record = value as Record<string, unknown>;
    if (typeof record.name === "string") {
      return normalizeWhitespace(record.name);
    }
  }
  return undefined;
}

function extractBrandName(schema: Record<string, unknown> | null): string | undefined {
  if (!schema) {
    return undefined;
  }
  const brand = schema.brand;
  if (typeof brand === "string") {
    return normalizeWhitespace(brand);
  }
  if (brand && typeof brand === "object") {
    const record = brand as Record<string, unknown>;
    if (typeof record.name === "string") {
      return normalizeWhitespace(record.name);
    }
  }
  return undefined;
}

function extractBreadcrumbCategory($: cheerio.CheerioAPI): string | undefined {
  const items = $("nav[aria-label*='breadcrumb'] a, .breadcrumb a")
    .toArray()
    .map((element) => normalizeWhitespace($(element).text()))
    .filter(Boolean);

  return items.length >= 2 ? items[items.length - 2] : undefined;
}

function buildAliases(productName: string, brandName: string | undefined, manualAliases: string[]): string[] {
  const candidates = [productName];
  if (brandName && productName.toLowerCase().startsWith(brandName.toLowerCase())) {
    candidates.push(productName.slice(brandName.length).trim().replace(/^[-\s]+/, ""));
  }
  candidates.push(productName.replace(/-/g, " "));
  candidates.push(...manualAliases);
  return uniquePreserveOrder(candidates);
}

function firstNonEmpty(...values: Array<string | null | undefined>): string | undefined {
  for (const value of values) {
    const cleaned = normalizeWhitespace(value);
    if (cleaned) {
      return cleaned;
    }
  }
  return undefined;
}

function tokenizeForMatch(value: string): string[] {
  const normalized = normalizeWhitespace(value)
    .normalize("NFKD")
    .replace(/[\u0300-\u036f]/g, "")
    .toLowerCase()
    .replace(/[^a-z0-9\s-]/g, " ")
    .replace(/\s+/g, " ")
    .trim();

  if (!normalized) {
    return [];
  }

  return normalized.split(" ").filter((token) => token.length > 1);
}

function inferBrandFromProductName(productName: string): string | undefined {
  const cleaned = normalizeWhitespace(productName);
  if (!cleaned) {
    return undefined;
  }

  const firstToken = cleaned.split(" ")[0];
  if (!firstToken) {
    return undefined;
  }

  // Ignore purely numeric tokens.
  if (/^\d+$/.test(firstToken)) {
    return undefined;
  }

  return firstToken;
}
